### PR TITLE
Bloodsucker Colors

### PR DESCRIPTION
**Pull request by Greniza**
## About The Pull Request

Bloodsucker HUDs now depict vassal and master as the same color- a purple-outlined vampire icon is the master of the purple vassals, for example. There are eight colors, randomly assigned, so no extra meta-knowledge about how many bloodsuckers there are.

image
_A coven of eight vampires, with the vampires closest to their coffins, and their favorite vassals closest to the aisle. None have broken the Masquerade._

image
_Blue has not broken the Masquerade, for now._

image
_But her name is known to the curator! She has broken the masquerade!_

If more than eight bloodsuckers (somehow) appear, then the ones following the first eight will each share a color with one of the first eight. The second set of eight will all have different colors from each other, as will the third set of eight, and the fourth set of eight, et cetera. I figure that this is acceptable because the only way that I can see more than eight appearing in any given round is admin shenanigans.

## Why It's Good For The Game

It allows bloodsuckers to determine at a glance whose vassal is whose, without needing to examine.

## Changelog

:cl:
qol: bloodsucker HUDs for vassals and masters are now color-coordinated
imageadd: bloodsucker favorite vassal icons are now their own sprite, rather than being a palette-swap of the regular vassal sprite.
/:cl:

### CODE REVIEW

**Comment by Greniza** [comment]
This PR still needs to test vassalization, as only sentient mobs can be vassalized.

**Comment by SpaceDragon00** [comment]
Why test your code? I just bully tom into bussing it and see what happens

**Comment by JohnFulpWillard** [comment]
Current huds are actually broken because of TG’s antag hud remake. I’d rather you fix it then given everyone theirs own Hud because having a different hud icon is supposed to immediately let you know something is off, like that they broke the masquerade, or are the favorite vassal.

**Comment by Greniza** [comment]
> Having a different hud icon is supposed to immediately let you know something is off, like that they broke the masquerade, or are the favorite vassal.

I was careful to preserve this in my current setup - The main focus of the Bloodsucker HUD icon is still the old face, so ones that broke the Masquerade remain obvious, and the Favorite Vassal sprites are different ([I laid them out here](https://discord.com/channels/279359975250198528/624445063879065620/943546444550381638))

If I encounter any issues with the HUDs not working as intended, I'll do my best to deal with them.

**Comment by Sheits** [comment]
In my opinion, the fav vassal should have a different coloured HUD too, as someone who hasn't played bloodsucker enough to memorise the symbols might not be able to tell between them

**Comment by Greniza** [comment]
@Sheits I'd post screenshots to argue that the regular and favorite vassal icons are distinct enough to be able to tell the difference at a glance but it's really hard to get more than one vassal with how you can't vassalize non-sentient or SSD beings.

**Comment by Sheits** [comment]
> @Sheits I'd post screenshots to argue that the regular and favorite vassal icons are distinct enough to be able to tell the difference at a glance but it's really hard to get more than one vassal with how you can't vassalize non-sentient or SSD beings.

As someone whos had over 10 vassals in one department at the same time i'd imagine its quite difficult to find the favourited one in the middle of a fight, especially being an antag that emphasises stealth so they could be changing outfits/getting more gear, i do like how the colours have turned out though

File: fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm [hunk 1]
```diff
@@ -726,14 +755,14 @@
...
-	add_team_hud(owner.current)
+	base_hud_name = "masquerade_broken"
+	update_vassal_hud(owner.current)
 	for(var/datum/mind/clan_minds as anything in get_antag_minds(/datum/antagonist/bloodsucker))
 		if(owner == clan_minds)
 			continue
 		if(!isliving(clan_minds.current))
 			continue
-		to_chat(clan_minds, span_userdanger("[owner.current] has broken the Masquerade! Ensure they are eliminated at all costs!"))
+		to_chat(clan_minds, span_userdanger("[owner.current], the [vassal_color]-colored bloodsucker, has broken the Masquerade! Ensure they are eliminated at all costs!"))
```
Comment (Greniza) [inline fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm:765]: If this isn't a change that you want to see, let me know.

File: fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm [hunk 2]
```diff
@@ -718,6 +725,28 @@
...
+		for(var/datum/antagonist/bloodsucker/individual_bloodsuckers in mind_members.antag_datums)
+			if(mind_members.has_antag_datum(/datum/antagonist/vassal)) // Favorite vassals do not have individual colors
+				continue
+			PossibleColors.Remove(individual_bloodsuckers.vassal_color)
+		
+		if(PossibleColors.len == 0)
+			PossibleColors = initial(PossibleColors)
+	
+	vassal_color = pick(PossibleColors)
+	update_vassal_hud(owner.current)
```
Comment (SpaceDragon00) [inline fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm:743]: Whys it called vassal hud if it's for /datum/antagonist/bloodsucker

**Comment by SpaceDragon00** [comment]
Im still of the belief that the masquerade/broken masquerade icons simply arent different enough for their size. It's such a small distinction that I feel would be extremely easy to miss

File: fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm (hunk 2 above)
Comment (Greniza) [reply fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm:743]: HUD depicting and designating vassals- different from things which depict things like daylight and blood count, which are also referred to as a HUD.
update_hud is afaik also already a thing.

File: fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm [hunk 3]
```diff
@@ -24,6 +24,10 @@
 	var/bloodsucker_title
 	///Used for assigning your reputation
 	var/bloodsucker_reputation
+	///HUD color
+	var/vassal_color
+	///antag_hud_name is dynamic because of colors. Update THIS, not antag_hud_name.
+	var/base_hud_name = "bloodsucker"
```
Comment (SpaceDragon00) [inline fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm:30]: This could probably just be a define

**Comment by Greniza** [comment]
> Im still of the belief that the masquerade/broken masquerade icons simply arent different enough for their size. It's such a small distinction that I feel would be extremely easy to miss

good call, going to invert the colors inside the eye.

File: fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm (hunk 2 above)
Comment (Greniza) [reply fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm:743]: depicts vassals.
bloodsucker hud is already a thing, the blood and Sol overlay, and afaik `update_hud` is also something, that I don't want to overwrite.

File: fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm (hunk 3 above)
Comment (Greniza) [reply fulp_modules/features/antagonists/bloodsuckers/code/antagonists/bloodsucker_datum.dm:30]: base_hud_name gets changed when masquarade gets broken.

**Comment by Greniza** [comment]
better @SpaceDragon00?

**Comment by JohnFulpWillard** [comment]
I can’t even figure out which hud is which, I have a hard time believing new players would know either.

**Comment by Greniza** [comment]
> I can’t even figure out which hud is which, I have a hard time believing new players would know either.

Between the vassal/bloodsucker, or between the individual ranks?

**Comment by SpaceDragon00** [comment]
> I can’t even figure out which hud is which, I have a hard time believing new players would know either.

No HUD lets new players know what their purpose is simply by their design. If a new player is seeing this HUD, they can always examine, since the examine text often tells you if someone is BS/Vassal/Otherwise if you can see the HUD. What's important is that the HUDs are thematic and distinct, more than anything

In the end, "Head Revolutionary" is a blue R and "Revolutionary" is a red R. There is nothing connecting these symbols and what they represent except for the knowledge you have gained through experience

**Comment by SpaceDragon00** [comment]
It would also be a good idea to evaluate how your icons look to people with different forms of colorblindness. You inevitably can't account for ALL of them (Especially monochromacy), but it might help some players if you could check and see if there's any quick fixes that don't muddy the presentation too much, since SS13 doesn't have a dedicated color-blindness mode

https://www.color-blindness.com/coblis-color-blindness-simulator/

**Comment by SpaceDragon00** [comment]
Also, as for the new broken masquerade icon, I'm okay with it, as the difference is a lot more noticeable now

**Comment by Greniza** [comment]
> > > I can’t even figure out which hud is which, I have a hard time believing new players would know either.
> > 
> > 
> > No HUD lets new players know what their purpose is simply by their design. If a new player is seeing this HUD, they can always examine, since the examine text often tells you if someone is BS/Vassal/Otherwise if you can see the HUD. What's important is that the HUDs are thematic and distinct, more than anything
> > In the end, "Head Revolutionary" is a blue R and "Revolutionary" is a red R. There is nothing connecting these symbols and what they represent except for the knowledge you have gained through experience
> 
> https://github.com/fulpstation/fulpstation/blob/3a973f4df86c5b1058fa08ee06651fa164e043f0/code/modules/antagonists/revolution/revolution.dm#L59
> 
> It’s also pretty easy to understand ‘R’ mean a Revolution in this case.

In all fairness, Bloodsuckers lacked anything differentiating icons from each other being expressly told to the player before this PR. I'll work on it.

**Comment by JohnFulpWillard** [comment]
> > > > I can’t even figure out which hud is which, I have a hard time believing new players would know either.
> > > 
> > > 
> > > No HUD lets new players know what their purpose is simply by their design. If a new player is seeing this HUD, they can always examine, since the examine text often tells you if someone is BS/Vassal/Otherwise if you can see the HUD. What's important is that the HUDs are thematic and distinct, more than anything
> > > In the end, "Head Revolutionary" is a blue R and "Revolutionary" is a red R. There is nothing connecting these symbols and what they represent except for the knowledge you have gained through experience
> > 
> > 
> > https://github.com/fulpstation/fulpstation/blob/3a973f4df86c5b1058fa08ee06651fa164e043f0/code/modules/antagonists/revolution/revolution.dm#L59
> > 
> > It’s also pretty easy to understand ‘R’ mean a Revolution in this case.
> 
> In all fairness, Bloodsuckers lacked anything differentiating icons from each other being expressly told to the player before this PR. I'll work on it.

E5B6AD08-1433-4AE9-8FE4-52740108AE22

**Comment by Greniza** [comment]
> > > > > I can’t even figure out which hud is which, I have a hard time believing new players would know either.
> > > > 
> > > > 
> > > > No HUD lets new players know what their purpose is simply by their design. If a new player is seeing this HUD, they can always examine, since the examine text often tells you if someone is BS/Vassal/Otherwise if you can see the HUD. What's important is that the HUDs are thematic and distinct, more than anything
> > > > In the end, "Head Revolutionary" is a blue R and "Revolutionary" is a red R. There is nothing connecting these symbols and what they represent except for the knowledge you have gained through experience
> > > 
> > > 
> > > https://github.com/fulpstation/fulpstation/blob/3a973f4df86c5b1058fa08ee06651fa164e043f0/code/modules/antagonists/revolution/revolution.dm#L59
> > > 
> > > It’s also pretty easy to understand ‘R’ mean a Revolution in this case.
> > 
> > 
> > In all fairness, Bloodsuckers lacked anything differentiating icons from each other being expressly told to the player before this PR. I'll work on it.
> 
> E5B6AD08-1433-4AE9-8FE4-52740108AE22

I said I'd work on it.

**Comment by Greniza** [comment]
merge conflict is fake afaik, resolved that this morning

**Comment by Helianthus00** [comment]
I don't really understand why this is needed.

**Pull request closed without merging**